fusemanager/fusemanager.go

```diff
@@ -0,0 +1,257 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+/*
+   Copyright 2019 The Go Authors. All rights reserved.
+   Use of this source code is governed by a BSD-style
+   license that can be found in the NOTICE.md file.
+*/
+
+package fusemanager
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/containerd/containerd/log"
+	"github.com/moby/sys/mountinfo"
+	"github.com/pkg/errors"
+	bolt "go.etcd.io/bbolt"
+
+	pb "github.com/containerd/stargz-snapshotter/fusemanager/api"
+	"github.com/containerd/stargz-snapshotter/service"
+	"github.com/containerd/stargz-snapshotter/snapshot"
+)
+
+const (
+	fuseManagerReady = iota
+	fuseManagerNotReady
+)
+
+type Server struct {
+	pb.UnimplementedFileSystemServiceServer
+
+	lock  sync.RWMutex
+	ready int32
+
+	// root is the latest root passed from containerd-stargz-grpc
+	root string
+	// config is the latest config passed from containerd-stargz-grpc
+	config *service.Config
+	// fsMap stores all filesystems from various configs
+	fsMap sync.Map
+	// curFs is filesystem created by latest config
+	curFs snapshot.FileSystem
+	ms    *bolt.DB
+}
+
+func NewFuseManager(ctx context.Context, fuseStoreAddr string) (*Server, error) {
+	if err := os.MkdirAll(filepath.Dir(fuseStoreAddr), 0700); err != nil {
+		return nil, errors.Wrapf(err, "failed to create directory %q", filepath.Dir(fuseStoreAddr))
+	}
+
+	db, err := bolt.Open(fuseStoreAddr, 0666, &bolt.Options{Timeout: 10 * time.Second, ReadOnly: false})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to configure fusestore")
+	}
+
+	fm := &Server{
+		ready: fuseManagerNotReady,
+		lock:  sync.RWMutex{},
+		fsMap: sync.Map{},
+		ms:    db,
+	}
+
+	return fm, nil
+}
+
+func (fm *Server) Init(ctx context.Context, req *pb.InitRequest) (*pb.Response, error) {
+	fm.lock.Lock()
+	fm.ready = fuseManagerNotReady
+	defer func() {
+		fm.ready = fuseManagerReady
+		fm.lock.Unlock()
+	}()
+
+	ctx = log.WithLogger(ctx, log.G(ctx))
+
+	config := &service.Config{}
+	err := json.Unmarshal(req.Config, config)
+	if err != nil {
+		log.G(ctx).WithError(err).Errorf("failed to get config")
+		return &pb.Response{}, err
+	}
+	fm.root = req.Root
+	fm.config = config
+
+	fs, err := service.NewFileSystem(ctx, fm.root, fm.config)
+	if err != nil {
+		return &pb.Response{}, err
+	}
+	fm.curFs = fs
+
+	err = fm.restoreFuseInfo(ctx)
```

Comment:
This will mount new filesystems even on the already-mounted mountpoints.
Restore should be performed at the first time Init is called.

Reply:
`restoreFuseInfo` calls `mount`, and `mount` will [skip layers](https://github.com/ilyee/stargz-snapshotter/blob/fuse-manager/fusemanager/fusemanager.go#L245) whose mountpoint is in `fsMap`, which means layer is mounted.

But here I doubt is it more proper to move this 'skip' to `restoreFuseInfo` rather than in `mount`? btw I'll add more comments for `restoreFuseInfo`

Reply:
@ilyee nvm about it, sorry. misunderstood about `fsMap` .